test_rclcpp/test/test_services_client.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <test_rclcpp/srv/add_two_ints.hpp>
+
+TEST(test_services_client, test_add)
+{
+  auto node = rclcpp::Node::make_shared("test_services_client");
+
+  auto client = node->create_client<test_rclcpp::srv::AddTwoInts>("add_two_ints");
+  auto request = std::make_shared<test_rclcpp::srv::AddTwoInts::Request>();
+  request->a = 1;
+  request->b = 2;
+
+  auto result = client->async_send_request(request);
+
+  rclcpp::spin_until_future_complete(node, result);  // Wait for the result.
+
+  EXPECT_EQ(3, result.get()->sum);
+}
+
+TEST(test_services_client, test_async_maintains_order)
+{
+  rclcpp::WallRate rate(30);
+
+  rclcpp::executors::SingleThreadedExecutor executor;
+
+  // Test that two consecutive requests are processed in the correct order
+  auto node = rclcpp::Node::make_shared("test_services_client");
+
+  auto client = node->create_client<test_rclcpp::srv::AddTwoInts>("add_two_ints");
+  auto request1 = std::make_shared<test_rclcpp::srv::AddTwoInts::Request>();
+  request1->a = 7;
+  request1->b = 8;
+
+  auto request2 = std::make_shared<test_rclcpp::srv::AddTwoInts::Request>();
+  request2->a = 9;
+  request2->b = 10;
+
+  auto result1 = client->async_send_request(request1);
+
+  // NOTE(esteve): for some reason, we need to sleep for a short time to ensure that
+  // messages are sent to the server
+  rate.sleep();
```

Comment:
Is one of the requests getting lost when sent without a sleep in between? If yes, we should probably create a ticket for that since that should be possible, right?